### ISSUE TITLE
pix2pix - Testing on a new image

### ISSUE THREAD

**Issue by s4chin** [tdeboissiere/DeepLearningImplementations#29]
The validation images generated during training appear to be in much better quality than the same images using the saved model(Using the saved model from the same epoch). I think it is due to some error on my testing code. I would be glad if you could help me out here - 
```py
    from scipy.misc import imsave
    img_path = os.path.join('path/to/input/image.jpg')
    img = cv2.imread(img_path)
    img = img[:, :, ::-1]  # GBR to RGB
    data = normalization(np.expand_dims(img, 0).transpose(0, 3, 1, 2)).transpose(0, 2, 3, 1)
    out = generator_model.predict(data) # generator_model is already loaded with weights
    out = inverse_normalization(out)
    print out.shape # (256, 256, 3)
    out=out[0, :, :, :]
    print out.shape # (256, 256, 3)
    imsave('path/to/output/image.jpg', out)
```
I think it has something to do with `.asarray(np.float32)` and `np.uint8`.

**Comment by tdeboissiere** [comment]
Which keras version are you using ?
It could have something to do with the Batch Norm mode (in early versions, you have a mode to compute BN statistics for inference in the same way as for training, i.e. use a per batch mean and standard dev. If you are in this mode, then when you make predictions, the estimated BN statistics given only one image may not be very good)

Do you have image samples ?
Do you get the same quality if you save with cv2.imwrite ?

**Comment by s4chin** [comment]
Keras version is 1.2.2
For using `cv2.imwrite` i have to multiply out by 255, but the saved images look more or less the same.
Here are the images(at 350th epoch)  generated by the `plot_generated_batch`-
![current_batch_350_100validation](https://cloud.githubusercontent.com/assets/5997735/24918560/6faa8adc-1efe-11e7-8f88-9c0417fff191.png)
Here are the same generated images at same epoch by the above code -
![0079](https://cloud.githubusercontent.com/assets/5997735/24918646/b9b1eb02-1efe-11e7-89f6-80295349601f.jpg)
![0059](https://cloud.githubusercontent.com/assets/5997735/24918665/c6b9707c-1efe-11e7-827e-9d2d68f61bba.jpg)
![0102](https://cloud.githubusercontent.com/assets/5997735/24918679/d1913c8c-1efe-11e7-8b26-f6f12356c4e2.jpg)
![0088](https://cloud.githubusercontent.com/assets/5997735/24918691/dc5d2568-1efe-11e7-875a-457ce75291b1.jpg)
Also, I now run the above code every 50 epochs during training itself using the generator_model so as to remove any errors that may be caused due to saving and loading models.

**Comment by tdeboissiere** [comment]
If I'm not mistaken, during training, predictions are made in batch before being saved:

`
X_pred = generator.predict(X_batch)  # X_batch is maybe (16,3,256,256)
for i in range(X_pred.shape[0]): cv2.imwrite(X_pred[i])
`

However you call `predict` on a single image.

2 possibilities:

- You are in the BN mode where moving mean and stddev are computed at training to be re-used at test time. In this case I would expect the same results whether you predict on a single image of a whole batch of image.
- You are in the BN mode where moving mean and stddev are computed at test time for each new batch you pass. You will get different results depending on how many images are in your test batch. That may explain your artifacts.


Did you try running your image saving code on multiple images at the same time (i.e. data has more than 1 sample on axis=0) ? Do check the mode argument in the BatchNorm layer.

I met a few float/int issues before but these led to images far more deteriorated than the ones you show.

**Comment by s4chin** [comment]
Alright, it was the case 2. Predicting in a batch led to no artifacts, and patches. I predicted on multiple images together and it worked. :+1: 
Also, having a batch of 4 gave slightly better results in terms of colors than a batch of 16. I guess that has something to do with having a batch_size of 4 during training too.
Thank you!
